### PR TITLE
Refactored checkpointing to use torch's distributed checkpointing package.

### PR DESCRIPTION
**Pull request by adil-a**
- Changed `save_utils.py` to use [torch.distributed.checkpoint](https://pytorch.org/docs/stable/distributed.checkpoint.html#)
- Added forward prefetching in `model_utils.py`
- Slightly modified Trainer class to support new changes

### CODE REVIEW

File: vectorlm/utils/save_utils.py [hunk 1]
```diff
@@ -171,13 +143,13 @@ def save_consolidated_model(
             torch.save(state_dict, save_path)
 
 
-def save_optimizer(
+def save_model_optimizer(
```
Comment (jwilles) [inline vectorlm/utils/save_utils.py]: A bit of a nitpick but can we name this something that is more expressive "save_model_and_optimizer" maybe?

File: vectorlm/utils/save_utils.py [hunk 2]
```diff
@@ -187,51 +159,67 @@ def save_optimizer(
...
+            storage_writer=writer,
+            process_group=model.process_group,
+            planner=DefaultSavePlanner(),
+        )
+        if rank == 0:
+            print(f"States saved to {output_dir}")
 
 
-def load_optimizer(
+def load_model_optimizer(
```
Comment (jwilles) [inline vectorlm/utils/save_utils.py]: Same comment as above. Not sure if this is clear that it loads both the model and optimizer.

File: vectorlm/utils/model_utils.py [hunk 3]
```diff
@@ -194,6 +194,7 @@ def _module_init_fn(module: nn.Module) -> Callable:
     ret_dict["auto_wrap_policy"] = auto_wrap_policy
     ret_dict["sharding_strategy"] = sharding_strategy
     ret_dict["device_id"] = torch.cuda.current_device()
+    ret_dict["forward_prefetch"] = True
```
Comment (jwilles) [inline vectorlm/utils/model_utils.py:197]: Is this related to the FSDP profiling?

File: vectorlm/trainer.py [hunk 4]
```diff
@@ -149,8 +147,8 @@ def save_checkpoint(self, epoch: int) -> None:
         )
         if rank == 0:
             save_metadata(save_dir, meta_dict)
-        save_model(self.model, save_dir, rank)
-        save_optimizer(self.optimizer, self.model, save_dir, rank)
+        save_model_optimizer(self.optimizer, self.model, save_dir, rank)
+        dist.barrier()
```
Comment (jwilles) [inline vectorlm/trainer.py]: Do we need two barriers here?

File: vectorlm/utils/save_utils.py [hunk 5]
```diff
@@ -270,9 +259,33 @@ def load_scheduler(
         rank: The worker's rank.
 
     """
-    sched_name = f"scheduler_rank{rank}.bin"
+    sched_name = "scheduler.bin"
     input_scheduler_file = os.path.join(input_dir, sched_name)
-    print(f"Loading scheduler state from {input_scheduler_file}")
+    if rank == 0:
```
Comment (jwilles) [inline vectorlm/utils/save_utils.py:265]: We should think about a master rank logger if we are using these conditionals for logging. We don't need to address it now.

File: vectorlm/trainer.py (hunk 4 above)
Comment (adil-a) [reply vectorlm/trainer.py]: Yes, there's a race condition otherwise in HSDP since `save_model_optimizer` immediately returns on replicated groups. save_scheduler expects a directory to already be made, but I can make it cleaner by just ensuring os.makedirs is called inside save_scheduler too.

File: vectorlm/utils/model_utils.py (hunk 3 above)
Comment (adil-a) [reply vectorlm/utils/model_utils.py:197]: Yes, but not entirely. The CUDA syncs that we saw in Jacob's table turned out to be time spent of global CUDA syncs. Regardless, this option should allow for full communication overlap in the forward pass.
